File: Design-Patterns/TemplateMethod/Cpp/circular_damage_spell_validator.cpp
#include "spell_casting_validator.h"
#include "game.h"

circular_damage_spell_validator::circular_damage_spell_validator(std::shared_ptr<character> character, std::shared_ptr<spell> spell) : spell_casting_validator(character, spell)
{
	
}

bool circular_damage_spell_validator::is_resource_enouth()
{
	if (character_->get_health() >= 10 && character_->get_mana() >= 90)
	{
		return true;
	}
	else
	{
		return false;
	}
}

bool circular_damage_spell_validator::is_skill_cooling()
{
	if (spell_->get_cool_down_counter() > 0)
	{
		return true;
	}
	else
	{
		return false;
	}
}

bool circular_damage_spell_validator::is_any_target_in_attack_range()
{
	std::unique_ptr<character> target = game::ray_cast_character(character_->position, 5);

	if (target == nullptr)
	{
		return false;
	}
	else
	{
		return true;
	}
}

File: Design-Patterns/TemplateMethod/CSharp/SpellCastingValidators/CircularDamageSpellValidator.cs
namespace TemplateMethod.SpellCastingValidators
{
    public class CircularDamageSpellValidator : SpellCastingValidator
    {
        public CircularDamageSpellValidator(Character character, Spell spell) : base(character, spell)
        {

        }

        protected override bool IsResourceEnough()
        {
            if (Character.Health >= 10 && Character.Mana >= 90)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        protected override bool IsSkillCooling()
        {
            if (Spell.CoolDownCounter > 0)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        protected override bool IsAnyTargetInAttackRange()
        {
            Character target = Game.RayCastCharacter(Character.Position, 5);

            if (target == null)
            {
                return false;
            }
            else
            {
                return true;
            }
        }
    }
}
